rust/wasmtime_bindings/src/memory_creator.rs

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+use anyhow::{anyhow, Result};
+use std::{cmp, ptr, slice, u32};
+use wasmtime::LinearMemory;
+
+const WASM_PAGE_SIZE: usize = 64 * 1024;
+
+extern "C" {
+    fn aligned_alloc(align: usize, size: usize) -> *mut u8;
+    fn free(ptr: *mut u8);
+}
+
+pub struct ScyllaLinearMemory {
+    ptr: *mut u8,
+    size: usize,
+    maximum_size: Option<usize>,
+}
+
+// The entire ScyllaLinearMemory is used only in a single thread,
+// because we're not sharing it between seastar shards
+unsafe impl Send for ScyllaLinearMemory {}
+unsafe impl Sync for ScyllaLinearMemory {}
+
+impl Drop for ScyllaLinearMemory {
+    fn drop(&mut self) {
+        // previously allocated or reset to nullptr in grow_to()
+        unsafe { free(self.ptr) };
+    }
+}
+
+unsafe impl LinearMemory for ScyllaLinearMemory {
+    fn byte_size(&self) -> usize {
+        self.size
+    }
+    fn maximum_byte_size(&self) -> Option<usize> {
+        self.maximum_size
+    }
+    fn grow_to(&mut self, new_size: usize) -> Result<()> {
+        let new_size_aligned = (new_size + WASM_PAGE_SIZE - 1) & !(WASM_PAGE_SIZE - 1);
```

Comment:
What is the maximum value of the `new_size` parameter? If it can be close to `usize::MAX` then the expression for `new_size_aligned` can overflow to 0. Perhaps we should calculate `(new_size + WASM_PAGE_SIZE - 1)` using the `checked_add` function (calculate `new_size.checked_add(WASM_PAGE_SIZE - 1)` and return Err if the result was None).

Reply:
I don't see any external factors limiting the `new_size` parameter, but even though it makes little sense to use it with a value close to `usize::MAX`, I agree that it's better to be consistent and use the `checked_add` method for this.